server/core/storage_v2.go

```diff
@@ -0,0 +1,623 @@
+	"github.com/pingcap/failpoint"
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/log"
+	"github.com/tikv/pd/pkg/encryption"
+	"github.com/tikv/pd/pkg/errs"
+	"github.com/tikv/pd/server/encryptionkm"
+	"github.com/tikv/pd/server/kv"
+	"go.etcd.io/etcd/clientv3"
+)
+
+// StorageV2 is the interface for the backend storage of the PD.
+// V2 is just a temp name, we will rename it to Storage eventually.
+// TODO: replace the core.Storage with this interface later.
+type StorageV2 interface {
+	ConfigStorage
+	MetaStorage
+	RuleStorage
+	ComponentStorage
+	ReplicationStatusStorage
+	GCSafePointStorage
+}
+
+type defaultStorage struct {
+	kv.Base
+	encryptionKeyManager *encryptionkm.KeyManager
+}
+
+// ConfigStorage defines the storage operations on the config.
+type ConfigStorage interface {
+	LoadConfig(cfg interface{}) (bool, error)
+	SaveConfig(cfg interface{}) error
+	LoadAllScheduleConfig() ([]string, []string, error)
+	SaveScheduleConfig(scheduleName string, data []byte) error
+	RemoveScheduleConfig(scheduleName string) error
+}
+
+// LoadConfig loads config from configPath then unmarshal it to cfg.
+func (s *defaultStorage) LoadConfig(cfg interface{}) (bool, error) {
+	value, err := s.Load(configPath)
+	if err != nil {
+		return false, err
+	}
+	if value == "" {
+		return false, nil
+	}
+	err = json.Unmarshal([]byte(value), cfg)
+	if err != nil {
+		return false, errs.ErrJSONUnmarshal.Wrap(err).GenWithStackByCause()
+	}
+	return true, nil
+}
+
+// SaveConfig stores marshallable cfg to the configPath.
+func (s *defaultStorage) SaveConfig(cfg interface{}) error {
+	value, err := json.Marshal(cfg)
+	if err != nil {
+		return errs.ErrJSONMarshal.Wrap(err).GenWithStackByCause()
+	}
+	return s.Save(configPath, string(value))
+}
+
+// LoadAllScheduleConfig loads all schedulers' config.
+func (s *defaultStorage) LoadAllScheduleConfig() ([]string, []string, error) {
+	prefix := customScheduleConfigPath + "/"
+	keys, values, err := s.LoadRange(prefix, clientv3.GetPrefixRangeEnd(prefix), 1000)
+	for i, key := range keys {
+		keys[i] = strings.TrimPrefix(key, prefix)
+	}
+	return keys, values, err
+}
+
+// SaveScheduleConfig saves the config of scheduler.
+func (s *defaultStorage) SaveScheduleConfig(scheduleName string, data []byte) error {
+	configPath := path.Join(customScheduleConfigPath, scheduleName)
+	return s.Save(configPath, string(data))
+}
+
+// RemoveScheduleConfig removes the config of scheduler.
+func (s *defaultStorage) RemoveScheduleConfig(scheduleName string) error {
+	configPath := path.Join(customScheduleConfigPath, scheduleName)
+	return s.Remove(configPath)
+}
+
+// MetaStorage defines the storage operations on the PD cluster meta info.
+type MetaStorage interface {
+	LoadMeta(meta *metapb.Cluster) (bool, error)
+	SaveMeta(meta *metapb.Cluster) error
+	LoadStore(storeID uint64, store *metapb.Store) (bool, error)
+	SaveStore(store *metapb.Store) error
+	SaveStoreWeight(storeID uint64, leader, region float64) error
```

Comment:
Can we use SaveStore to replace it?

Reply:
It seems that they are saving different things. So having two methods is reasonable to me.

Reply:
It just clones a store and uses SaveStore which I think we don't need in the interface.